src/core/server/elasticsearch/errors.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Boom from 'boom';
+import { get } from 'lodash';
+
+const code = Symbol('ElasticsearchError');
+
+const CODE_NOT_AUTHORIZED = 'Elasticsearch/notAuthorized';
+
+export interface ElasticsearchError extends Boom {
+  [code]?: string;
+}
+
+function isElasticsearchError(error: any): error is ElasticsearchError {
+  return Boolean(error && error[code]);
+}
+
+function decorate(
+  error: Error,
+  errorCode: string,
+  statusCode: number,
+  message?: string
+): ElasticsearchError {
+  if (isElasticsearchError(error)) {
+    return error;
+  }
+
+  const boom = Boom.boomify(error, {
+    statusCode,
+    message,
+    override: false,
+  }) as ElasticsearchError;
+
+  boom[code] = errorCode;
+
+  return boom;
+}
+
+/**
+ * Helpers provided to simplify future migration away from Boom as internal Error.
```

Comment:
I think this is a really good idea, but maybe the documentation should be more forceful so that we know people will adopt it and won't look inside the error object.

```suggestion
 * Helpers for working with errors returned from the Elasticsearch service. Since the internal data of errors are subject to change, consumers of the Elasticsearch service should always use these helpers to classify errors instead of checking error internals such as `body.error.header[WWW-Authenticate]`
```